### PR TITLE
geotagging: extends date before unix epoch -> 01/01/01 00:00:00

### PR DESCRIPTION
**Pull request by phweyland**
as discussed on [darktable-dev](https://www.mail-archive.com/darktable-dev@lists.darktable.org/msg06481.html).

Use GDateTime instead of unix time_t.

### CODE REVIEW

File: src/libs/geotagging.c [hunk 1]
```diff
@@ -170,11 +170,12 @@ static void _apply_offset_callback(GtkWidget *widget, dt_lib_module_t *self)
...
 {
   dt_lib_geotagging_t *d = (dt_lib_geotagging_t *)self->data;
-  if(d->datetime > 0)
+  if(d->datetime)
   {
+    char *dt = g_date_time_format(d->datetime, "%Y:%m:%d %H:%M:%S");
     char text[DT_DATETIME_LENGTH];
-    struct tm dt;
-    strftime(text, sizeof(text), "%Y:%m:%d %H:%M:%S", localtime_r(&d->datetime, &dt));
+    memcpy(&text, dt, sizeof(text));
```
Comment (TurboGit) [inline src/libs/geotagging.c:177]: use g_strlcpy as memcpy() won't add the \0 if we overflow text.

Comment (TurboGit) [reply src/libs/geotagging.c:177]: And maybe just use dt in dt_control_datetime() directly and remove text?

File: src/libs/geotagging.c [hunk 2]
```diff
@@ -1276,82 +1275,75 @@ static void _display_datetime(dt_lib_datetime_t *dtw, const time_t datetime,
...
+    _new_datetime(datetime, self);
   }
-  else return 0;
 }
 
-static time_t _get_image_datetime(dt_lib_module_t *self)
+static GDateTime *_get_datetime_from_text(const char *text, GTimeZone *tz)
+{
+  char datetime_s[DT_DATETIME_LENGTH];
+  memcpy(&datetime_s, text, sizeof(datetime_s));
```
Comment (TurboGit) [inline src/libs/geotagging.c:1324]: g_strlcpy to ensure \0 at end.

**Review by TurboGit — CHANGES_REQUESTED** [review]
A quick review, I'll test tomorrow.

**Comment by phweyland** [comment]
Thanks ! All done.

File: src/libs/geotagging.c [hunk 3]
```diff
@@ -1276,82 +1273,75 @@ static void _display_datetime(dt_lib_datetime_t *dtw, const time_t datetime,
...
-  // if difference display the new date
-  if(datetime != -1 &&
-     (dt.tm_year != dt2.tm_year || dt.tm_mon != dt2.tm_mon ||  dt.tm_mday != dt2.tm_mday ||
-      dt.tm_hour != dt2.tm_hour || dt.tm_min != dt2.tm_min || dt.tm_sec != dt2.tm_sec))
-  {
-    _display_datetime(&d->dt, datetime, TRUE, self);
-  }
-  return datetime;
+
+  const int year = atol(gtk_entry_get_text(GTK_ENTRY(d->dt.widget[0])));
```
Comment (TurboGit) [inline src/libs/geotagging.c:1280]: any reason to use `atol()` to fill `int` vars?

**Review by TurboGit — COMMENTED** [review]
A minor comment, apart from that it is good to go. Thanks.

File: src/libs/geotagging.c (hunk 3 above)
Comment (phweyland) [reply src/libs/geotagging.c:1280]: I haven't any good reason, really. Thanks for pinpointing this.

**Comment by phweyland** [comment]
@TurboGit , I've just checked that time_t is also used sometimes in the import process, [import from camera] and [copy & import], to read exif time. Do you prefer I fix this in this PR or in an other one ? (I think I could do that today).

**Review by TurboGit — APPROVED** [review]
Looks perfect to me now! Thanks.

**Comment by TurboGit** [comment]
Another PR please.

**Pull request merged**